xilem_core/src/views/adapt.rs

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use core::marker::PhantomData;
+
+use crate::{DynMessage, MessageResult, Mut, View, ViewId, ViewPathTracker};
+
+/// A view that wraps a child view and modifies the state that callbacks have access to.
+pub struct Adapt<
+    ParentState,
+    ParentAction,
+    ChildState,
+    ChildAction,
+    Context,
+    ChildView,
+    ProxyFn = fn(
+        &mut ParentState,
+        AdaptThunk<ChildState, ChildAction, Context, ChildView>,
+    ) -> MessageResult<ParentAction>,
+> where
+    Context: ViewPathTracker,
+{
+    proxy_fn: ProxyFn,
+    child: ChildView,
+    #[allow(clippy::type_complexity)]
+    phantom: PhantomData<fn() -> (ParentState, ParentAction, ChildState, ChildAction, Context)>,
+}
+
+/// A "thunk" which dispatches an message to an adapt node's child.
+///
+/// The closure passed to [`Adapt`] should call this thunk with the child's
+/// app state.
+pub struct AdaptThunk<'a, ChildState, ChildAction, Context, ChildView>
+where
+    Context: ViewPathTracker,
+    ChildView: View<ChildState, ChildAction, Context>,
+{
+    child: &'a ChildView,
+    view_state: &'a mut ChildView::ViewState,
+    id_path: &'a [ViewId],
+    message: DynMessage,
+}
+
+/// A view that wraps a child view and modifies the state that callbacks have access to.
+///
+/// # Examples
+///
+/// Suppose you have an outer type that looks like
+///
+/// ```ignore
+/// struct TodoList {
+///     todos: Vec<Todo>
+/// }
+/// ```
+///
+/// and an inner type/view that looks like
+///
+/// ```ignore
+/// struct Todo {
+///     label: String
+/// }
+///
+/// struct TodoView {
+///     label: String
+/// }
+///
+/// enum TodoAction {
+///     Delete
+/// }
+///
+/// impl<ViewCtx> View<Todo, TodoAction, ViewCtx> for TodoView {
+///     // ...
+/// }
+/// ```
+///
+/// then your top-level action (`()`) and state type (`TodoList`) don't match `TodoView`'s.
+/// You can use the `Adapt` view to mediate between them:
+///
+/// ```ignore
+/// state
+///     .todos
+///     .enumerate()
+///     .map(|(idx, todo)| adapt(
+///         TodoView { label: todo.label },
+///         |data: &mut AppState, thunk| {
+///             thunk.call(&mut data.todos[idx]).map(|action| match action {
+///                 TodoAction::Delete => data.todos.remove(idx),
+///             })
+///         })
+///     )
+/// ```
+pub fn adapt<ParentState, ParentAction, ChildState, ChildAction, Context, ChildView, ProxyFn>(
+    child: ChildView,
+    proxy_fn: ProxyFn,
+) -> Adapt<ParentState, ParentAction, ChildState, ChildAction, Context, ChildView, ProxyFn>
+where
+    ChildState: 'static,
+    ChildAction: 'static,
+    ParentState: 'static,
+    ParentAction: 'static,
+    Context: ViewPathTracker + 'static,
+    ChildView: View<ChildState, ChildAction, Context>,
+    ProxyFn: Fn(
+            &mut ParentState,
+            AdaptThunk<ChildState, ChildAction, Context, ChildView>,
+        ) -> MessageResult<ParentAction>
+        + 'static,
```

Comment:
All those bounds add up a lot of noise (one reason why I wanted to get rid of the `'static` bound), but I think it's more correct to provide the same bounds here as for the `View` implementation to avoid misuse...